discovery/linode/linode.go

```diff
@@ -0,0 +1,255 @@
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	TagSeparator    string         `yaml:"tag_separator,omitempty"`
+}
+
+// Name returns the name of the Config.
+func (*SDConfig) Name() string { return "linode" }
+
+// NewDiscoverer returns a Discoverer for the Config.
+func (c *SDConfig) NewDiscoverer(opts discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(c, opts.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery periodically performs Linode requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client       *linodego.Client
+	port         int
+	tagSeparator string
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "linode_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client := linodego.NewClient(
+		&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		},
+	)
+	client.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version))
+	d.client = &client
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"linode",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "Linode",
+	}
+
+	// Gather all linode instances.
+	instances, err := d.client.ListInstances(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Gather detailed IP address info for all IPs on all linode instances.
+	detailedIPs, err := d.client.ListIPAddresses(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range instances {
+		if len(instance.IPv4) == 0 {
+			continue
+		}
+
+		var (
+			privateIPv4, publicIPv4, publicIPv6, backupsStatus string
+			extraIPs                                           []string
+		)
+
+		for _, detailedIP := range detailedIPs {
+			if detailedIP.LinodeID != instance.ID {
+				continue
+			}
+
+			if detailedIP.Type == "ipv6" {
+				publicIPv6 = detailedIP.Address
+			} else if detailedIP.Type == "ipv4" {
+				// Linode's /networking/ips endpoint does not return the IPs sorted. To ensure we
```

Comment:
Can't we rely on the order of the /instance endpoint?

Reply:
We can, I was trying to avoid iterating over the `detailedIPs` slice multiple times. But the IPs returned from the `/instances` endpoint _are_ sorted, so we can definitely lean on that for consistent ordering. I'll play around with it 👍